walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,63 @@
 
 package com.google.sps;
 
-import java.util.Collection;
+import java.util.*;
 
 public final class FindMeetingQuery {
+  /**
+   * Returns the times when {@code request} could happen on a day containing {@code events}
+   * Time complexity of the algorithm is O(nlogn + mn), in which
+   * n is the number of {@code events}, and m is the number of attendees in {@code request}
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+    Collection<TimeRange> availableTimes = new ArrayList<>();
+    List<Event> eventList = new ArrayList<>(events);
+    eventList.sort(Comparator.comparingInt(e -> e.getWhen().start()));
+
+    // Locate gaps between events joined by any of the request attendees
+    int prevEndTime = TimeRange.START_OF_DAY;
+    for(Event e: eventList) {
+      // Ignore if event not attended by any of the request attendees
+      if(!attendeesOverlap(e, request)) {
+        continue;
+      }
+      int eventStartTime = e.getWhen().start();
+      int eventEndTime = e.getWhen().end();
+
+      // Check if event overlaps with preceding events
+      if(eventStartTime <= prevEndTime) {
+        prevEndTime = Math.max(eventEndTime, prevEndTime);
+      } else {
+        int timeGap = eventStartTime - prevEndTime;
+        // Check if gap between events can hold the meeting duration
+        if(timeGap >= request.getDuration()) {
+          availableTimes.add(TimeRange.fromStartDuration(prevEndTime, timeGap));
+        }
+        prevEndTime = eventEndTime;
+      }
+    }
+
+    // Check for the gap after the last event
```

Comment:
Can this potentially be refactored into a smaller function?

Reply:
Refactored into new `checkGap` method.